src/Core/PolynomialFunction.java

```diff
@@ -129,31 +132,98 @@ public void derive() {
         functionString =  sb.toString();
     }
 
-    public void calcRoots(double start, double end, double step) {
-        roots = new ArrayList<>();
-        Vector2D prevPoint = evaluate(start);
+    public void calcRootsAndExtremes(double start, double end, double step) {
+        // If function is a constant skip the calculation
+        if(coefficients.length == 1) {
+            roots = new HashSet<>();
+            extremePoints = new HashSet<>();
+            roots.add(Double.NaN);
+            extremePoints.add(Double.NaN);
+            return;
+        }
+        Callable<Set<Double>> task1 = () -> calcRoots(start, end, step);
+        Callable<Set<Double>> task2 = () -> calcExtremes(start, end, step);
+
+        try {
+            List<Future<Set<Double>>> results = executorService.invokeAll(Arrays.asList(task1, task2));
+            roots = results.get(0).get();
+            extremePoints = results.get(1).get();
+        } catch (InterruptedException | ExecutionException e) {
+            e.printStackTrace();
+        }
+    }
+
+    private static final double EPSILON = 1E-15;
+
+    private Set<Double> calcRoots(double start, double end, double step) {
+        Set<Double> roots = new LinkedHashSet<>();
+        double prevY = evaluate(start).y;
         for (double x = start + step; x <= end; x += step) {
-            Vector2D point = evaluate(x);
-            if (prevPoint.y * point.y <= 0) {
-                roots.add(x - step / 2);
+            double y = evaluate(x).y;
+            // If the value is incredibly close to zero, it is a root => Add
+            if (Math.abs(y) < EPSILON) {
+                double root = formatDouble(x);
+                if (shouldAddValue(root, roots)) {
+                    roots.add(root);
+                }
+            // First use Incremental Search to check for a sign change (if-statement)
+            // then calculate an approximately really close position using linear interpolation
+            } else if (prevY * y < 0) {
+                double root = (x - step) + (x - (x - step)) * Math.abs(prevY) / (Math.abs(prevY) + Math.abs(y));
```

Comment:
isn't (x - (x - step)) just step?

Reply:
yeah i think you are right i will test it. I just wrote it stupidly after a linear interpolation formula i found and it worked :D

Reply:
Done